core/src/main/java/com/linecorp/armeria/client/WebSocketHttp1ResponseWrapper.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import java.util.concurrent.TimeUnit;
+
+import com.linecorp.armeria.common.ClosedSessionException;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.internal.client.ClientRequestContextExtension;
+import com.linecorp.armeria.internal.client.DecodedHttpResponse;
+
+import io.netty.channel.EventLoop;
+
+final class WebSocketHttp1ResponseWrapper extends AbstractHttpResponseWrapper {
+
+    WebSocketHttp1ResponseWrapper(DecodedHttpResponse delegate,
+                                  EventLoop eventLoop, ClientRequestContext ctx,
+                                  long responseTimeoutMillis, long maxContentLength) {
+        super(delegate, eventLoop, ctx, responseTimeoutMillis, maxContentLength);
+        final ClientRequestContextExtension extension = ctx.as(ClientRequestContextExtension.class);
+        if (extension != null) {
+            extension.setClosingResponseTask(cause -> {
+                close0(cause, false);
+            });
+        }
+    }
+
+    @Override
+    void close(@Nullable Throwable cause, boolean cancel) {
+        if (cancel || !(cause instanceof ClosedSessionException)) {
+            close0(cause, cancel);
+            return;
+        }
+        // Do not close the response right away to give a chance to WebSocketFrameDecoder to close the
+        // response normally if it receives a close frame.
```

Comment:
I've been staring at this piece of code for 30min just to remember why `ClosedSessionException` was excluded 😅 

```suggestion
        // response normally if it receives a close frame instead of completing
        // exceptionally due to a ClosedSessionException.
```